### DISCUSSION TITLE
Implement custom OAuth and control route access

### DISCUSSION THREAD

**Discussion by marcantoinegodde** [fastapi/fastapi#8677] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
I don't have any code.
```

### Description

Hello there,

I'm discovering FastAPI. I'm desperately trying to implement the OAuth of my association to protect specific routes. I really spent a lot of time to look for a solution on the Internet, without success. It is a very standard OAuth 2.0 without OpenID (exactly like the Github one https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps).
I can do the implementation myself, create the middleware, but I was totally unable to find something to store session easily. With Express, I just had to import express-session with a Redis DB, I could create/access variables very easily. It seems there's nothing to do similar things. Should I go with JWT instead of session to store `state` variable for instance ?
As I use FastAPI, it would be nice to use the auto-generated doc with the "padlock" stuff.

In summary, with what should I go to implement an OAuth ?
A huge thanks in advance.

### Operating System

macOS

### FastAPI Version

0.70.0

### Python Version

Python 3.9.7

**Comment by mayteio** [comment]
FWIW you can use dependencies to check auth status.

```python
class OAuth2PasswordBearerCookie(OAuth2):
  def __init__(self, tokenUrl, scopes):
    flows = OAuthFlowsModel(password={"tokenUrl": tokenUrl, "scopes": scopes})
    super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

  async def __call__(self, request):
    header_authorization = request.headers.get("Authorization")
    # ...

oauth2_scheme = OAuth2PasswordBearerCookie(tokenUrl="/api/auth/token", auto_error=False)

async def is_admin(
    token: str = Depends(oauth2_scheme)
):
  # do things to check token, raise if not authorized

@app.route("/", dependencies = Depends(is_admin))
async def get_protected_route():
  # do route things
```
